### PR TITLE
fix FMA modes for ILS Approach when disconnecting AP. Also add new fl…

### LINKED ISSUE #146: AP disconnect change FMA and APPR mode when established on ILS
Happened in the last dev build 

When established on an ILS and disconnecting the autopilot with a keybind the FMA change from SPEED to THR LVR, and G/S to OP DES and on the FCU The APPR mode is disengaged and LOC is engaged

Here is a video of it 

https://youtu.be/BBap9r_d0pE

### LINKED ISSUE #1222: [BUG] ILS FMA changes when disconnecting AP
**Mod Version**

Latest master

**Describe the bug**

After intercepting the G/S and LOC with the AP on, I disconnected the AP. The FMA reverted to "CLB" instead of remaining on the glideslope so I had to press the "APPR" button again on the FDU to recapture. The autothrust mode also changed to "THR CLB" when it should remain in "SPEED" mode. The FMA/FD should stay in "G/S LOC" when disconnecting the AP.

**To Reproduce**

1. With AP on, hit APPR button to intercept the localizer and glideslope.
2.Once established on the ILS approach, disconnect the autopilot (using the red button on the sidestick).
3.Note the FMA change (THR CLB, CLB) and pitch flight director no longer is giving glideslope guidance.

**Expected behavior**

When disconnecting the AP on an ILS approach, the FMA/FD should stay in "G/S LOC" so that you can use the flight directors to hand-fly the ILS. The autothrust mode should also remain in "SPEED" mode.

**Actual behavior**

Plane correctly stays in "LOC"  but the pitch mode reverts to "CLB" which renders the pitch FD useless. The autothrust mode also switched to "THR CLB."

**References**

This was just after disconnecting the AP on the ILS 16C into KSEA:
Screenshot (93)

Here is V1 Simulations (Real airbus captain) disconnecting the AP on an ILS in the FFA320. Note the plane stays in (G/S LOC) and the autothrust stays in (SPEED):
https://youtu.be/FsROrvv9tJs?t=6847

**Additional context**

Was this working before/when did the issue start occurring?

Is this a problem in the vanilla unmodded game?

Discord username (if different from GitHub):

### LINKED ISSUE #1705: GS disappears on approach when AP off
**Mod Version**
built: 2020-11-01T01:01:54+00:00
ref: refs/heads/master
sha: af858b4cd6c37cf087a04545ff1aa6c2b931f3b7
actor: Benjozork
event_name: push

**Describe the bug**
On an ILS with APPR engaged and AP on i have LOC and GS modes active.  When I disconnect the AP to land, aircraft goes into CLB and THR CLB on FMA.  This should still be GS/Speed as the active mode.  The thrust did not change, but it seemed to be only an indication on the FMA.

**To Reproduce**
Fly an ILS approach and engage APPR.  Disconnect autopilot and look at the FMA.

**Expected behavior**
When AP off, should stay SPEED/GS/LOC on FMA

**Actual behavior**
When I diconnected the AP, THR CLB and CLB show on FMA

**References**
A-320 pilot 

Was this working before/when did the issue start occurring?
unsure

Is this a problem in the vanilla unmodded game?
unsure
<

### PR DESCRIPTION
**Pull request by antikythera288**
Fixes #1222 
Fixes #1705
Fixes #146

## Summary of Changes

- [x] Many fixes on the several possible FMA modes.
- [x] Implementation of Goaround (including MCDU Pages and AP Modes).

Limitations:

- Re-engaging the AP with earlier armed APPR will only succeed when you are on, - or above - the glide slope (check magenta diamonds). If you are below, OP CLB will be initiated until you "De-" and Re-Activate the APPR button. This is probably a problem with how the Asobo code works internally. The GS HOLD event is driven somewhere internally and its really a mess.

- During enganged APPR Mode, you should be able to pre-select a HDG for goaround. Right now you can pre-select it (validate on PFD) but its still "dashed" (---) on the FCU

- As we dont have yet missed approach flight plans or something similar, during GOAROUND you will see HDG armed with GA TRK. The GA TRK is implemented with the WING LEVEL Event and can be exited when pulling the HDG knob or exits automatically when plane reaches thrust reduction altitude, thus enganging HDG to the synchronized current heading.

- When flying the GOAROUND by hand, the FMA's will show correct modes for it, except the FD's ... This seems to be always wrong when you disable the AP (vanilla bug) ... 

## Screenshots (if necessary)

will upload video soon from APPR Mode to GOAROUND and LAND

## References

- Help from IRL Pilot (@St54Kevin)

- http://www.a320dp.com/

- https://www.theairlinepilots.com/forumarchive/a320/a320-fma.pdf

- https://youtu.be/FsROrvv9tJs?t=6847

## Additional context

Discord username (if different from GitHub): mrmercury#8898

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

**Comment by Benjozork** [comment]
Added my modelbehavior changes. Needs to be tested.

**Comment by antikythera288** [comment]
> 
> 
> Added my modelbehavior changes. Needs to be tested.

it seems the template is not working ... LOC and APPR are "lighted up" and are not clickeable (not even Toolip or Cursor)

grafik

**Comment by Benjozork** [comment]
I will take a look

**Comment by Benjozork** [comment]
🤦 I forgot to include the file

**Comment by Benjozork** [comment]
Fixed it. Current implementation has one bug I noticed, where sometimes the APPR button can be in the wrong state and turning it OFF actually enables `AUTOPILOT GLIDESLOPE HOLD`.

**Comment by Oberstumfuher** [comment]
I do not know what is happening. I'm using dev mod, built: 2020-11-03T14: 27: 20 + 00: 00, everything correct, takeoff, cruise FL380, approach by ILS, is correctly in the center of the airstrip, however when making the autoland the aircraft looks that does not activate the brakes because it is jumping, like a ping-pong ball, for 3 or 4 times until it loses speed, the brakes were at maximum, VLS 125nm, could anyone check what may be happening?

**Comment by antikythera288** [comment]
> 
> 
> Fixed it. Current implementation has one bug I noticed, where sometimes the APPR button can be in the wrong state and turning it OFF actually enables `AUTOPILOT GLIDESLOPE HOLD`. I'll see what I can do.

will come online soon to do some testing and further FMA enhancements ... so regarding your mentioned bug, i guess the only code place which is related to this is the change in A32NX_Interior_Autopilot.xml right?

**Comment by Benjozork** [comment]
@Oberstumfuher that is unrelated to this PR.

**Comment by antikythera288** [comment]
major push pending, will test later together with one of our IRL Pilots for proper implementation

**Comment by tareksabet** [comment]
can you add it fixes issue #146

**Comment by Benjozork** [comment]
@tareksabet added it

**Comment by antikythera288** [comment]
@St54Kevin you are now cleared to final testing ;) ...

File: A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_Autopilot.xml
Comment (Benjozork) [inline A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_Autopilot.xml:8]: We should probably call this `FBW_Airbus_Autopilot_Master`.

File: A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_Autopilot.xml
Comment (Benjozork) [inline A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_Autopilot.xml:27]: We can remove the surrounding `Switch` and `Case` tags, since there is only ever 2 APs on the A320.

File: A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_Autopilot.xml
Comment (Benjozork) [inline A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_Autopilot.xml:55]: Same here.

File: A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_Misc.xml
Comment (Benjozork) [inline A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_Misc.xml:249]: You can probably just revert this file to what it is in master, no ? I think you told me about this being broken

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_MainDisplay.js
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_MainDisplay.js:403]: Remove commented code

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_MainDisplay.js
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_MainDisplay.js:1216]: Remove `else if` block to avoid unnecessarily evaluating those expressions.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js:2040]: Remove `console.log` calls for performance.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js:2100]: Same here.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js:2125]: Same here.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js:2132]: Remove `else if` block to avoid unnecessarily evaluating those expressions.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/PFD/Airbus_FMA.js
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/PFD/Airbus_FMA.js:1559]: Remove commented code.

**Review by Benjozork — CHANGES_REQUESTED** [review]
Great work and congrats !

File: A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_Autopilot.xml
Comment (antikythera288) [reply A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_Autopilot.xml:27]: this is more or less a copy of the existing asobo template, except the LEFT_SINGLE_CODE Part. As you know this RNP/Template stuff is still my week point. I tried to remove everything were i was sure i can :P ... So in this case ... do i really only remove the 4 lines and its fine? Because if i just can delete them and the syntax remains intact, i can do this aswell for all other switch/case in this template (there are about 2-3 others)

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_MainDisplay.js
Comment (antikythera288) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_MainDisplay.js:403]: this i would like to keep together with the comment ... i still dont understand why asobo had this code there in the first place ... maybe i miss something now and i need to comment it in again? If we all agree its nonsense, i delete it though ...

File: A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_Misc.xml
Comment (antikythera288) [reply A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_Misc.xml:200]: reverted to master

File: A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_Autopilot.xml
Comment (antikythera288) [reply A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_Autopilot.xml]: agree

File: A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_Autopilot.xml
Comment (antikythera288) [reply A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_Autopilot.xml]: can you please review if the new commit i fine ? i tested inflight and it seems i did not brake something at least :P

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js
Comment (antikythera288) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js:2124]: will remove when testing was successfull... is there any label for that existing? ... its crucial for debugging right now and i would like to keep it

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js
Comment (antikythera288) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js:2040]: will remove when testing was successfull... is there any label for that existing? ... its crucial for debugging right now and i would like to keep it

File: A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_Autopilot.xml
Comment (Benjozork) [reply A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_Autopilot.xml]: LGTM

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js
Comment (Benjozork) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js:2040]: Okay it'll have to be removed before merge

Comment (antikythera288) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js:2040]: ok, tomorrow our IRL Pilots will have a further look and then i will remove the console.log's

**Comment by St54Kevin** [comment]
Category: Accuracy Check
Name: St54Kevin #3583
Date of testing: Current date 12/11/2020
Version of the sim: 1715 PR pull via installer
PR Tested: #1715 
**Object of test:** Approach modes and FMA
- did a short flight from EDDW to EDDH, ILS RW15. route all flown in NAV including STAR and TRANS. 

**Expected IRL Behavior:**

**Current PR behavior:**
Approach was perfect. Disconnected AP at 1000ft via the AP Disc Push button on the joystick. Flight director did NOT hold GS for me after disconnecting AP, although GS in active mode. I Went TOGA at 200ft. All FMA modes were as expected. **FMA:** MAN TOGA, SRS (OP CLB blue), GA TRK (HDG blue), A/THR blue. Flaps one notch up, gear up. THR RED ALT went perfect. Levers to CLB. **FMA:** THR CLB, OP CLB (ALT blue), NAV. Here the plane now tries to turn back to the last flight plan waypoint which was the RW15. I expected the plane to switch to HDG mode (it was armed according to the FMA, see above).


Flight notes (If required):
EDDW-EDDH, RW27-ILS15
ISA weather
all in NAV mode
GA without AP

Overall Rating: mostly acurate
Conclusions of the testing: If HDG is armed, it should switch to HDG not to NAV.

Will do other testes with AP ON and with Vectors to Approach etc.

**Comment by Benjozork** [comment]
Testing phase is started by maintainers.

**Review by Benjozork — APPROVED** [review]

**Review by Carter633 — APPROVED** [review]
Name: British Avgeek UTube -- Carter633 #1436
Date of testing: Current date 13/11/2020
Version of the sim: 1.10.11.0
PR Tested: #1715 
Object of test: FMA Modes for ILS Approach and Go Around

Testing technique: Quick flight from EGHH up to EGGD Bristol using ILS Runway27 + Go-Around
Steps you've made to test the Pull Request:
1. Started C&D at EGHH - departed and headed up to join the BRI1C arrival via Compton
2. Vectored myself in to intercept the ILS, AP flew aircraft down to 200ft - FMA modes looked good
3. Executed Go-Around at 200ft
4. Flew as per the charts including tracking and overflying an NDB - AP behaved throughout.
5. Flew second approach - all normal, FMA modes still working well

Issues (If Any): None
Overall Rating: Good
Conclusions of the testing: Works great!

**Comment by Benjozork** [comment]
@antikythera288 are you going to address @St54Kevin's issue in this PR ? if not, I will merge it.

**Comment by antikythera288** [comment]
> @antikythera288 are you going to address @St54Kevin's issue in this PR ? if not, I will merge it.

Its already addressed and pushed some minutes after his complain 🤣😜

**Comment by Benjozork** [comment]
ah, didn't see your commit message :+1:

**Comment by jooglestumpe** [comment]
I'm not sure this is really an improvement to be honest. The FMA modes may display what they should but it still doesn't match what is actually happening. For example, when disengaging the AP, APPR remains engaged and G/S is still displayed, but the flight director doesn't want to follow the G/S and displays what it did before (in LOC mode). This can be somewhat misleading IMO.

**Comment by antikythera288** [comment]
> 
> 
> I'm not sure this is really an improvement to be honest. The FMA modes may display what they should but it still doesn't match what is actually happening. For example, when disengaging the AP, APPR remains engaged and G/S is still displayed, but the flight director doesn't want to follow the G/S and displays what it did before (in LOC mode). This can be somewhat misleading IMO.

the FMA Modes implementation is correct according to what we have available as definitions of every flight phase and other variables. So the conditions are valid actually for most ... What you point out is what i have written in the 4th line of the limitations in the PR description. If AP is OFF, the FD's are not following anything ... so its a general limitation we have. Also this PR introduces GOAROUND capability

**Comment by Carter633** [comment]
...and in due course the extra bits will be sorted with other PRs

**Pull request merged**
